spotify_tensorflow/luigi/python_dataflow_task.py

```diff
@@ -0,0 +1,303 @@
+
+    def on_successful_run(self):
+        """ Callback that gets called right after the dataflow job has finished successfully but
+        before validate_output is run.
+        """
+        pass
+
+    def validate_output(self):
+        """ Callback that can be used to validate your output before it is moved to it's final
+        location. Returning false here will cause the job to fail, and output to be removed instead
+        of published.
+
+        :return: Whether the output is valid or not
+        :rtype: Boolean
+        """
+        return True
+
+    def file_pattern(self):
+        """ If one/some of the input target files are not in the pattern of part-*,
+        we can add the key of the required target and the correct file pattern
+        that should be appended in the command line here. If the input target key is not found
+        in this dict, the file pattern will be assumed to be part-* for that target.
+
+        :return A dictionary of overrided file pattern that is not part-* for the inputs
+        :rtype: Dict of String to String
+        """
+        return {}
+
+    def run(self):
+        cmd_line = self._mk_cmd_line()
+        logger.info(" ".join(cmd_line))
+
+        try:
+            run_with_logging(cmd_line)
+        except subprocess.CalledProcessError as e:
+            logging.error(e, exc_info=True)
+            # exit luigi with the same exit code as the python dataflow job proccess
+            # In this way users can easily exit the job with code 50 to avoid Styx retries
+            # https://github.com/spotify/styx/blob/master/doc/design-overview.md#workflow-state-graph
+            os._exit(e.returncode)
+
+        self.on_successful_run()
+        if self.validate_output():
+            self._publish_outputs()
+        else:
+            raise ValueError("Output is not valid")
+
+    def _publish_outputs(self):
+        for (name, target) in self._output.items():
+            if hasattr(target, "publish"):
+                target.publish(self._output_uris[name])
+
+    def _mk_cmd_line(self):
+        cmd_line = self._dataflow_executable()
+        cmd_line.extend(self._get_dataflow_args())
+        cmd_line.extend(self._get_input_args())
+        cmd_line.extend(self._get_output_args())
+        cmd_line.extend(self.tfx_args())
+        cmd_line.extend(self.args())
+        return cmd_line
+
+    def _dataflow_executable(self):
+        """
+        Defines the executable used to run the python dataflow job.
+        """
+        return ["python", self.python_script]
+
+    def _get_input_uri(self, file_pattern, target):
+        uri = get_uri(target)
+        uri = uri.rstrip("/") + "/" + file_pattern
+        return uri
+
+    def _get_file_pattern(self):
+        file_pattern = self.file_pattern()
+        if not isinstance(file_pattern, dict):
+            raise ValueError("file_pattern() must return a dict type")
+        return file_pattern
+
+    def _get_input_args(self):
```

Comment:
I think this function is complex enough that a docstring would really help here.